common/src/main/java/io/netty/util/DomainWildcardMappingBuilder.java

```diff
@@ -0,0 +1,160 @@
+     * {@code null} values are forbidden for both hostnames and values.
+     * <p>
+     * <a href="https://tools.ietf.org/search/rfc6125#section-6.4">DNS wildcard</a> is supported as hostname. The
+     * wildcard will only match one sub-domain deep and only when wildcard is used as the most-left label.
+     *
+     * For example:
+     *
+     * <p>
+     *  *.netty.io will match xyz.netty.io but NOT abc.xyz.netty.io
+     * </p>
+     *
+     * @param hostname the host name (optionally wildcard)
+     * @param output   the output value that will be returned by {@link Mapping#map(Object)}
+     *                 when the specified host name matches the specified input host name
+     */
+    public DomainWildcardMappingBuilder<V> add(String hostname, V output) {
+        map.put(normalizeHostName(hostname),
+                checkNotNull(output, "output"));
+        return this;
+    }
+
+    private String normalizeHostName(String hostname) {
+        checkNotNull(hostname, "hostname");
+        if (hostname.isEmpty() || hostname.charAt(0) == '.') {
+            throw new IllegalArgumentException("Hostname '" + hostname + "'not valid");
+        }
+        hostname = ImmutableDomainWildcardMapping.normalize(checkNotNull(hostname, "hostname"));
+        if (hostname.charAt(0) == '*') {
+            if (hostname.length() < 3 || hostname.charAt(1) != '.') {
+                throw new IllegalArgumentException("Wildcard Hostname '" + hostname + "'not valid");
+            }
+            return hostname.substring(1);
+        }
+        return hostname;
+    }
+    /**
+     * Creates a new instance of an immutable {@link Mapping}.
+     *
+     * @return new {@link Mapping} instance
+     */
+    public Mapping<String, V> build() {
+        return new ImmutableDomainWildcardMapping<V>(defaultValue, new LinkedHashMap<String, V>(map));
+    }
+
+    private static final class ImmutableDomainWildcardMapping<V> implements Mapping<String, V> {
+        private static final String REPR_HEADER = "ImmutableDomainWildcardMapping(default: ";
+        private static final String REPR_MAP_OPENING = ", map: ";
+        private static final String REPR_MAP_CLOSING = ")";
+
+        private final V defaultValue;
+        private final Map<String, V> map;
+
+        ImmutableDomainWildcardMapping(V defaultValue, Map<String, V> map) {
+            this.defaultValue = defaultValue;
+            this.map = Collections.unmodifiableMap(map);
+        }
+
+        @Override
+        public V map(String hostname) {
+            if (hostname != null) {
+                hostname = normalize(hostname);
+
+                // Let's try an exact match first
+                V value = map.get(hostname);
+                if (value != null) {
+                    return value;
+                }
+
+                // No exact match, let's try a wildcard match.
+                int idx = hostname.indexOf('.');
```

Comment:
https://gist.github.com/apangin/7a9b7062a4bd0cd41fcc#file-hotspot-jvm-intrinsics-L66 : java 8 doesn't contain `String::indexOf(char)` in the list of `String` intrinsics, while the version using `String`, yes. Java 9 has introduced the version with `char` IIRC

Reply:
sure we can also use `String` if you think it matters a lot...  WDYT @franz1981 ?

Reply:
I've written a small bench to check if if it can really matter, will tell you soon :)

Reply:
Ok, for input Strings < 100, searching a `String` in whatever position is just slower then using `indexOf(char)`, so better to left `indexOf(char)`

Reply:
ok then I think I am done...